### DISCUSSION TITLE
fast-check integration

### DISCUSSION THREAD

**Discussion by krzysztofpniak** [cypress-io/cypress#24897] [Questions and Help]
I tried to integrate with fast-check, but unfortunately it requires promises which cypress does not provide. They are needed for counter example shrinking and enhancing thrown error by fast-check data like seed and path. As discussed with the author and he suggested: https://github.com/dubzzz/fast-check/issues/3408#issuecomment-1318232700

**Comment by portyu9** [comment]
Hi @krzysztofpniak 

For the `asyncProperty` / `asyncModelRun` use case, your diagnosis is essentially correct: Cypress commands and fast-check's asynchronous execution model do not compose directly.

Cypress commands such as:

```js
cy.get(...).type(...)
```

are **queued Cypress commands**, not normal Promises. Therefore this does not make the command awaitable:

```js
await cy.get(...).type(...)
```

Adding `async` to the fast-check command's `run()` method does not solve that mismatch either.

`fast-check` needs an asynchronous property invocation to actually finish before it can start the next run, detect the failure, and repeatedly execute smaller counterexamples during shrinking. Cypress, meanwhile, first builds its own command queue and executes that queue afterward.

For Cypress UI tests, one workable compromise is to use fast-check only to generate cases up front and then let Cypress execute those cases:

```js
const inputs = fc.sample(
  fc.string({ minLength: 1 }),
  { numRuns: 50, seed: 12345 }
)

inputs.forEach((text) => {
  cy.get('[data-testid="city-input"]')
    .clear()
    .type(text)

  // Cypress assertions...
})
```

That gives you generated/reproducible input data, but **not fast-check's automatic shrinking of failures**, because fast-check is no longer controlling each Cypress execution.

If shrinking/model-based testing is the important requirement, I would move that property test to code that returns real synchronous values or Promises (for example a unit/integration test under Mocha/Vitest/Jest) and keep Cypress for the browser-level scenarios.

Wrapping a Cypress chain in `Promise.resolve()`, `Cypress.Promise`, or simply adding `await` is not a reliable bridge between the two execution models.

Please mark this answer as accepted if it helped, thank you.
